components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java

```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("Approved scope(s) : " + OAuth2Util.buildScopeString(authzReqMsgCtx.getApprovedScope()));
+        }
+        return true;
+    }
+
+    private boolean validateAccessDelegation(OAuth2AuthorizeReqDTO authzReqDTO, OAuth2AuthorizeRespDTO authorizeRespDTO,
+                                             OAuthAuthzReqMessageContext authzReqMsgCtx,
+                                             ResponseTypeHandler authzHandler) throws IdentityOAuth2Exception {
+        boolean accessDelegationAuthzStatus = authzHandler.validateAccessDelegation(authzReqMsgCtx);
+        if (!accessDelegationAuthzStatus) {
+            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNAUTHORIZED_CLIENT, "Authorization Failure!");
+            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+            if (log.isDebugEnabled()) {
+                log.debug("User : " + authzReqDTO.getUser() +
+                        " doesn't have necessary rights to grant access to the resource(s) : " +
+                        OAuth2Util.buildScopeString(authzReqDTO.getScopes()));
             }
+            return false;
         }
+        return true;
+    }
 
-	try {
-	    // set the authorization request context to be used by downstream handlers. This is introduced as a fix for
-	    // IDENTITY-4111
-	    OAuth2Util.setAuthzRequestContext(authzReqMsgCtx);
-	    authorizeRespDTO = authzHandler.issue(authzReqMsgCtx);
-	} finally {
-	    // clears authorization request context
-	    OAuth2Util.clearAuthzRequestContext();
-	}
-	
-        return authorizeRespDTO;
+    private boolean validateClient(OAuth2AuthorizeReqDTO authzReqDTO, OAuth2AuthorizeRespDTO authorizeRespDTO,
+                                   OAuthAuthzReqMessageContext authzReqMsgCtx, ResponseTypeHandler authzHandler)
+            throws IdentityOAuth2Exception {
+        boolean isAuthorizedClient = authzHandler.isAuthorizedClient(authzReqMsgCtx);
+        if (!isAuthorizedClient) {
+            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNAUTHORIZED_CLIENT,
+                    "The authenticated client is not authorized to use this authorization grant type");
+            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+            if (log.isDebugEnabled()) {
+                log.debug("Client validation failed for user : " + authzReqDTO.getUser() +
+                        ", for client : " + authzReqDTO.getConsumerKey());
+            }
+            return false;
+        }
+        return true;
+    }
+
+    private OAuthAuthzReqMessageContext getoAuthAuthzReqMessageContext(OAuth2AuthorizeReqDTO authzReqDTO)
+            throws IdentityOAuth2Exception, InvalidOAuthClientException {
+        OAuthAuthzReqMessageContext authzReqMsgCtx;
+        authzReqMsgCtx = new OAuthAuthzReqMessageContext(authzReqDTO);
```

Comment:
Shall we change the authzReqMsgCtx variable name as -> authorizeRequestMessageContext to improve readability